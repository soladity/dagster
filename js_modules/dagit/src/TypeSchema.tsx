import * as React from "react";
import gql from "graphql-tag";
import styled from "styled-components";
import { Colors } from "@blueprintjs/core";
import TypeWithTooltip from "./TypeWithTooltip";
import { TypeSchemaFragment } from "./types/TypeSchemaFragment";
import { TypeInfoFragment } from "./types/TypeInfoFragment";

interface ITypeSchemaProps {
  type: TypeSchemaFragment;
}

function renderTypeRecursive(
  type: TypeInfoFragment,
  typeLookup: { [typeName: string]: TypeInfoFragment },
  indent: string = ""
): React.ReactElement<HTMLElement> {
  if (type.isDict && "fields" in type) {
    const innerIndent = indent + "  ";
    return (
      <>
        {`{`}
        {type.fields.map(fieldData => (
          <DictEntry key={fieldData.name}>
            <DictBlockComment
              indent={innerIndent}
              content={fieldData.description}
            />
            {innerIndent}
            <DictKey>{fieldData.name}</DictKey>
            {fieldData.isOptional && Optional}
            {`: `}
            {renderTypeRecursive(
              typeLookup[fieldData.type.name],
              typeLookup,
              innerIndent
            )}
          </DictEntry>
        ))}
        {`${indent}}`}
      </>
    );
  }
  if (type.isList) {
    const innerType = type.innerTypes[0].name;
    return (
      <>[{renderTypeRecursive(typeLookup[innerType], typeLookup, indent)}]</>
    );
  }
  if (type.isNullable) {
    const innerType = type.innerTypes[0].name;
    return (
      <>
        {renderTypeRecursive(typeLookup[innerType], typeLookup, indent)}
        {Optional}
      </>
    );
  }
  return <TypeWithTooltip type={type} />;
}

export default class TypeSchema extends React.Component<ITypeSchemaProps> {
  static fragments = {
    TypeSchemaFragment: gql`
      fragment TypeInfoFragment on Type {
        name
        description
        isDict
        isList
        isNullable
        innerTypes {
          name
        }
        ... on CompositeType {
          fields {
            name
            description
            type {
              name
            }
            isOptional
          }
        }
        ...TypeWithTooltipFragment
      }

      fragment TypeSchemaFragment on Type {
        ...TypeInfoFragment
        innerTypes {
          ...TypeInfoFragment
        }
      }
      ${TypeWithTooltip.fragments.TypeWithTooltipFragment}
    `
  };

  public render() {
    const { type } = this.props;

    const innerTypeLookup = {};
    for (const innerTypeData of type.innerTypes) {
      innerTypeLookup[innerTypeData.name] = innerTypeData;
    }

    return (
      <TypeSchemaContainer>
        <DictBlockComment content={type.description} indent="" />
        {renderTypeRecursive(type, innerTypeLookup)}
      </TypeSchemaContainer>
    );
  }
}

const TypeSchemaContainer = styled.code`
  color: ${Colors.GRAY3};
  display: block;
  white-space: pre-wrap;
  font-size: smaller;
  line-height: 18px;
`;

const DictEntry = styled.div``;

const DictKey = styled.span`
  color: ${Colors.BLACK};
`;

const DictComment = styled.div`
  /* This allows long comments to wrap as nice indented blocks, while
     copy/pasting as a single line with space-based indentation. */
  text-indent: -1.85em;
  padding-left: 1.85em;
  white-space: initial;
`;

const DictBlockComment = ({
  indent = "",
  content
}: {
  indent: string;
  content: string | null;
}) =>
  content != null ? (
    <DictComment>{`${indent.replace(
      / /g,
      "\u00A0"
    )}/* ${content} */`}</DictComment>
  ) : null;

const Optional = (
  <span style={{ fontWeight: 500, color: Colors.ORANGE2 }}>?</span>
);
