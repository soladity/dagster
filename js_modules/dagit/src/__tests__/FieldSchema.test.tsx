import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

import TypeSchema from "../TypeSchema";

it("renders given a basic type", () => {
  let intConfigData = {
    type: {
      name: "Int",
      description: "An int",
      typeAttributes: {
        isNamed: true
      },
      isDict: false,
      isList: false,
      isNullable: false,
      innerTypes: []
    }
  };
  const component = TestRenderer.create(
    <BrowserRouter>
      <TypeSchema type={intConfigData.type} />
    </BrowserRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it("renders given a complex type", () => {
  let complexConfigData = {
    type: {
      name: "Dict.4",
      description: "",
      isDict: true,
      typeAttributes: { isNamed: false },
      isList: false,
      isNullable: false,
      fields: [
        {
          name: "field_one",
          type: {
            name: "String"
          },
          isOptional: false
        },
        {
          name: "field_two",
          type: {
            name: "String"
          },
          isOptional: false
        },
        {
          name: "field_three",
          type: {
            name: "String"
          },
          isOptional: false
        },
        {
          name: "nested_field",
          type: {
            name: "Dict.3"
          },
          isOptional: false
        }
      ],
      innerTypes: [
        {
          name: "String",
          description: "",
          isDict: false,
          isList: false,
          typeAttributes: { isNamed: true },
          isNullable: false,
          innerTypes: []
        },
        {
          name: "Dict.3",
          description: "",
          isDict: true,
          isList: false,
          typeAttributes: { isNamed: false },
          isNullable: false,
          innerTypes: [
            {
              name: "String"
            },
            {
              name: "Int"
            },
            {
              name: "List.Nullable.Int"
            },
            {
              name: "Nullable.Int"
            }
          ],
          fields: [
            {
              name: "field_four_str",
              type: {
                name: "String"
              },
              isOptional: false
            },
            {
              name: "field_five_int",
              type: {
                name: "Int"
              },
              isOptional: false
            },
            {
              name: "field_six_nullable_int_list",
              type: {
                name: "List.Nullable.Int"
              },
              isOptional: true
            }
          ]
        },
        {
          name: "Int",
          description: "",
          isDict: false,
          isList: false,
          typeAttributes: { isNamed: true },
          isNullable: false,
          innerTypes: []
        },
        {
          name: "List.Nullable.Int",
          description: "",
          isDict: false,
          isList: true,
          typeAttributes: { isNamed: false },
          isNullable: false,
          innerTypes: [
            {
              name: "Nullable.Int"
            },
            {
              name: "Int"
            }
          ]
        },
        {
          name: "Nullable.Int",
          description: "",
          isDict: false,
          isList: false,
          typeAttributes: { isNamed: false },
          isNullable: true,
          innerTypes: [
            {
              name: "Int"
            }
          ]
        }
      ]
    }
  };
  const component = TestRenderer.create(
    <BrowserRouter>
      <TypeSchema type={complexConfigData.type} />
    </BrowserRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
