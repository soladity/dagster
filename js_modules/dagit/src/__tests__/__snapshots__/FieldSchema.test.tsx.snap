// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders given a basic type 1`] = `
<code
  className="sc-bwzfXH kJdBIT"
>
  <div
    className="sc-ifAKCX OCUpN"
  >
    /* An int */
  </div>
  <a
    href="/?typeExplorer=Int"
    onClick={[Function]}
  >
    <code
      className="sc-bdVaJa bGXATf"
    >
      Int
    </code>
  </a>
</code>
`;

exports[`renders given a complex type 1`] = `
<code
  className="sc-bwzfXH kJdBIT"
>
  <div
    className="sc-ifAKCX OCUpN"
  >
    /*  */
  </div>
  {
  <div
    className="sc-htpNat KtFsv"
  >
      
    <span
      className="sc-bxivhb bkkmIq"
    >
      field_one
    </span>
    : 
    <a
      href="/?typeExplorer=String"
      onClick={[Function]}
    >
      <code
        className="sc-bdVaJa bGXATf"
      >
        String
      </code>
    </a>
  </div>
  <div
    className="sc-htpNat KtFsv"
  >
      
    <span
      className="sc-bxivhb bkkmIq"
    >
      field_two
    </span>
    : 
    <a
      href="/?typeExplorer=String"
      onClick={[Function]}
    >
      <code
        className="sc-bdVaJa bGXATf"
      >
        String
      </code>
    </a>
  </div>
  <div
    className="sc-htpNat KtFsv"
  >
      
    <span
      className="sc-bxivhb bkkmIq"
    >
      field_three
    </span>
    : 
    <a
      href="/?typeExplorer=String"
      onClick={[Function]}
    >
      <code
        className="sc-bdVaJa bGXATf"
      >
        String
      </code>
    </a>
  </div>
  <div
    className="sc-htpNat KtFsv"
  >
      
    <span
      className="sc-bxivhb bkkmIq"
    >
      nested_field
    </span>
    : 
    {
    <div
      className="sc-htpNat KtFsv"
    >
          
      <span
        className="sc-bxivhb bkkmIq"
      >
        field_four_str
      </span>
      : 
      <a
        href="/?typeExplorer=String"
        onClick={[Function]}
      >
        <code
          className="sc-bdVaJa bGXATf"
        >
          String
        </code>
      </a>
    </div>
    <div
      className="sc-htpNat KtFsv"
    >
          
      <span
        className="sc-bxivhb bkkmIq"
      >
        field_five_int
      </span>
      : 
      <a
        href="/?typeExplorer=Int"
        onClick={[Function]}
      >
        <code
          className="sc-bdVaJa bGXATf"
        >
          Int
        </code>
      </a>
    </div>
    <div
      className="sc-htpNat KtFsv"
    >
          
      <span
        className="sc-bxivhb bkkmIq"
      >
        field_six_nullable_int_list
      </span>
      <span
        style={
          Object {
            "color": "#BF7326",
            "fontWeight": 500,
          }
        }
      >
        ?
      </span>
      : 
      [
      <a
        href="/?typeExplorer=Int"
        onClick={[Function]}
      >
        <code
          className="sc-bdVaJa bGXATf"
        >
          Int
        </code>
      </a>
      <span
        style={
          Object {
            "color": "#BF7326",
            "fontWeight": 500,
          }
        }
      >
        ?
      </span>
      ]
    </div>
      }
  </div>
  }
</code>
`;
